src/Microsoft.Diagnostics.Monitoring.Egress.AzureStorage/AzureBlobEgressEndpoint.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.Egress.AzureStorage
+{
+    internal class AzureBlobEgressEndpoint : IEgressEndpoint<AzureBlobEgressStreamOptions>
+    {
+        private readonly AzureBlobEgressEndpointOptions _settings;
+
+        public AzureBlobEgressEndpoint(AzureBlobEgressEndpointOptions settings)
+        {
+            _settings = settings;
+        }
+
+        public async Task<EgressResult> EgressAsync(
+            Func<CancellationToken, Task<Stream>> action,
+            string name,
+            AzureBlobEgressStreamOptions options,
+            CancellationToken token)
+        {
+            using var stream = await action(token);
+
+            return await Upload(stream, name, options, token);
+        }
+
+        public async Task<EgressResult> EgressAsync(
+            Func<Stream, CancellationToken, Task> action,
+            string name,
+            AzureBlobEgressStreamOptions options,
+            CancellationToken token)
+        {
+            using var stream = new MemoryStream();
+
+            await action(stream, token);
```

Comment:
This looks bad for perf/reliability. I assume this is keeping everything cached in memory and for a large diagnostic artifact we will probably OOM. (I also noticed we have a pre-existing similar issue in the GCDump path, that one could be equally problematic)

At some point we either need a Stream that supports concurrent Read/Write with a bounded buffer or we could find/make some alternate Azure Blob client upload implementation that supports a push model instead of a pull model.

Reply:
I had the similar thoughts about that too. The dotnet-monitor v-team discussed, in one of our sync ups, how the tool needs to respect environment constraints (e.g. don't eat too much memory because it'll either OOM or be evicted).

This was not something that I wanted to tackle in the initial changes to support egress and triggers, but something that can probably holistically address in the near future.

One idea we had was to implement a new stream type that acted as a MemoryStream but then dumps to a temporary file on disk when a threshold is hit so that we can maintain in-memory performance so long as the threshold isn't exceeded.

Reply:
You can do streaming writes to blobs. Call "OpenWriteAsync" on the BlockBlobClient to get back a writable Stream.